### PR TITLE
Purge old free transactions when MemoryPoolSize

### PR DESCRIPTION
**Pull request by belane**
This is an approach to filter MemPool when max is reached, dropping free Tx older than 20 blocks.
When `MemoryPoolSize` is reached purge oldest free transactions from MemPool and, if still above, apply the current filter.

https://github.com/neo-project/neo/pull/393 is a simpler approach.

### CODE REVIEW

File: neo/Ledger/Blockchain.cs [hunk 1]
```diff
@@ -356,18 +356,27 @@ private RelayResultReason OnNewTransaction(Transaction transaction)
...
-                    mem_pool.TryRemove(hash, out _);
+                UInt256[] oldfree = mem_pool.GetValuesBy(p => p.timestamp < DateTime.UtcNow.AddSeconds(-SecondsPerBlock * 20) && p.tx.NetworkFee == Fixed8.Zero)
+                    .Select(p => p.Hash)
+                    .ToArray();
+                foreach (UInt256 hash in oldfree)
+                    mem_pool.TryRemove(hash, out _);
+
+                if (mem_pool.Count > MemoryPoolSize)
+                {
+                    UInt256[] delete = mem_pool.Values.AsParallel()
```
Comment (jsolman) [inline neo/Ledger/Blockchain.cs]: this should not be happening on every transaction with a large mem_pool. This should be done in the code after persisting a block

Comment (jsolman) [reply neo/Ledger/Blockchain.cs]: See my PR as an example of where this should be done: https://github.com/neo-project/neo/pull/394

**Comment by erikzhang** [comment]
But the transactions that was dropped will soon be relayed again.

**Comment by belane** [comment]
Currently the pool does not guarantee that a dropped Tx will not enter again.

**Comment by erikzhang** [comment]
I rearranged the code.

**Comment by erikzhang** [comment]
The static `mem_pool` will break the actor model, I changed it to non-static.
@belane @jsolman Please review it again.

**Comment by belane** [comment]
Looks all fine, ready to merge.

**Comment by jsolman** [comment]
You can merge it I guess, but as I said, but see my comment; I don’t think this gaurantees the free transactions actually stay purged.

**Comment by shargon** [comment]
@jsolman  With this we are able to call `RemoveLowestFee` when want, and stamp the transaction with the reception date could help to the consensus nodes for choose the more spread transactions

Maybe we need to call this method on `TryAdd`

File: neo/Ledger/Blockchain.cs [hunk 2]
```diff
@@ -352,22 +351,18 @@ private RelayResultReason OnNewTransaction(Transaction transaction)
...
-                UInt256[] delete = mem_pool.Values.AsParallel()
-                    .OrderBy(p => p.NetworkFee / p.Size)
-                    .ThenBy(p => p.NetworkFee)
-                    .ThenBy(p => new BigInteger(p.Hash.ToArray()))
-                    .Take(mem_pool.Count - MemoryPoolSize)
-                    .Select(p => p.Hash)
-                    .ToArray();
-                foreach (UInt256 hash in delete)
-                    mem_pool.TryRemove(hash, out _);
+                mem_pool.RemoveOldFree(DateTime.UtcNow.AddSeconds(-SecondsPerBlock * 20));
```
Comment (shargon) [inline neo/Ledger/Blockchain.cs]: This should be moved to `TryAdd`

**Comment by erikzhang** [comment]
@shargon @jsolman Can I merge this PR?

File: neo/Ledger/MemPool.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        public IEnumerator<Transaction> GetEnumerator()
+        {
+            return mem_pool.Select(p => p.Value.Transaction).GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
```
Comment (shargon) [inline neo/Ledger/MemPool.cs]: I will test that this pointed to https://github.com/neo-project/neo/pull/392/files#diff-692a215303b3871f385ba1044c5fe5feR40 ?

Comment (shargon) [reply neo/Ledger/MemPool.cs]: PASS :)

**Comment by erikzhang** [comment]
Waiting for @jsolman to approve it.

File: neo/Ledger/MemPool.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        public void RemoveLowestFee(int count)
+        {
+            if (count <= 0) return;
+            if (count >= mem_pool.Count)
+            {
+                mem_pool.Clear();
```
Comment (jsolman) [inline neo/Ledger/MemPool.cs]: So it would have to get twice the size of the mem_pool for this to happen (and this resets in that case). Is this ever expected to happen?

File: neo/Ledger/MemPool.cs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            }
+            else
+            {
+                UInt256[] delete = mem_pool.AsParallel()
+                    .OrderBy(p => p.Value.Transaction.NetworkFee / p.Value.Transaction.Size)
+                    .ThenBy(p => p.Value.Transaction.NetworkFee)
+                    .ThenBy(p => new BigInteger(p.Key.ToArray()))
+                    .Take(count)
```
Comment (jsolman) [inline neo/Ledger/MemPool.cs]: How well does the compiler optimize this if this is running to remove 1 transaction repeatedly?

File: neo/Ledger/MemPool.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                mem_pool.TryRemove(hash, out _);
+        }
+
+        public bool TryAdd(UInt256 hash, Transaction tx)
+        {
+            mem_pool.TryAdd(hash, new PoolItem(tx));
+
+            if (mem_pool.Count > MemoryPoolSize)
+            {
+                RemoveOldFree(DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 20));
```
Comment (jsolman) [inline neo/Ledger/MemPool.cs]: Once the mempool doesn't have any transactions that are free, but the pool is full this will repeatedly be called, maybe we could keep track of the lowest fee in the pool to know whether this should be called every time at transaction is added in the case that the pool is full of transactions that all have > 0 fee.

File: neo/Ledger/Blockchain.cs (hunk 2 above)
Comment (belane) [reply neo/Ledger/Blockchain.cs]: Done

File: neo/Ledger/MemPool.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                    .ThenBy(p => new BigInteger(p.Key.ToArray()))
+                    .Take(count)
+                    .Select(p => p.Key)
+                    .ToArray();
+                foreach (UInt256 hash in delete)
+                    mem_pool.TryRemove(hash, out _);
+            }
+        }
+
+        public void RemoveOldFree(DateTime time)
```
[the remark this answers is not in the document]
Comment (belane) [reply neo/Ledger/MemPool.cs]: Currently, the pool management is individual for each node, so nothing prevents the same transactions from being received again. Maybe we should think about solutions for these cases.

File: neo/Ledger/MemPool.cs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        public bool TryRemove(UInt256 hash, out Transaction tx)
+        {
+            if (mem_pool.TryRemove(hash, out PoolItem item))
+            {
+                tx = item.Transaction;
+                return true;
+            }
+            else
```
[the remark this answers is not in the document]
Comment (belane) [reply neo/Ledger/MemPool.cs]: During compilation both options should generate the same IL code. I think this is more readable, but we can change, what is the pattern we are using now?

File: neo/Ledger/MemPool.cs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        public bool TryGetValue(UInt256 hash, out Transaction tx)
+        {
+            if (mem_pool.TryGetValue(hash, out PoolItem item))
+            {
+                tx = item.Transaction;
+                return true;
+            }
+            else
```
[the remark this answers is not in the document]
Comment (belane) [reply neo/Ledger/MemPool.cs]: same here

File: neo/Ledger/MemPool.cs (hunk 6 above)
Comment (shargon) [reply neo/Ledger/MemPool.cs]: We can store a flag of how many fee tx are inside for call this method or not

Comment (erikzhang) [reply neo/Ledger/MemPool.cs]: Or we can create two pools, one high priority and the other low priority.

Comment (jsolman) [reply neo/Ledger/MemPool.cs]: I like the solution with a separate data structure for the free transactions.

File: neo/Ledger/MemPool.cs [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        private class PoolItem
+        {
+            public readonly Transaction Transaction;
+            public readonly DateTime Timestamp;
+
+            public PoolItem(Transaction tx)
+            {
+                Transaction = tx;
+                Timestamp = DateTime.UtcNow;
```
Comment (vncoelho) [inline neo/Ledger/MemPool.cs]: As @erikzhang commented, transactions gonna expire and, perhaps, RPC seeds nodes will relay it again.
Malicious RPC nodes can even run a loop for relaying the same transaction because the DateTime.UtcNow is given by the time that each node receives, right @jsolman and @shargon ?

Another option (but far from our reality) would be to sign the time of the transaction...aehauheauheauea But I think it would be too much for now.

Comment (jsolman) [reply neo/Ledger/MemPool.cs]: Yes, I mentioned early along in this PR that this won’t prevent the transactions from getting back in the mem_pool. It can potentially help some old ones die out, but no guarantee even without someone purposely relaying them.

Free transactions will probably eventually need to have a proof of work to order them; even trying to sort them by some random lottery will still really amount to a proof of work since the algorithm will be known.

Free transactions that need to guarantee they aren’t in the mempool for too long should protect themselves from being stuck in the pool for too long by including something to make them invalid after a certain number of blocks. This can be done today already for invocation transactions. Enforcing Signing the time of all transactions by the sender really isn’t the right answer anyway for how to order free transactions, because an old transaction arguably should have higher priority than a newer one; it has been waiting longer after all and they are both free. Keep in mind if the transaction gets old it is because of a high amount of free transactions causing some to not be accepted due to the 20 free per block minimum. Having it get old doesn’t mean it is any less valid than a new one that just got relayed. So, as stated if the transaction should be become invalid after a period of not being relayed it is the responsibility of the creator of the transaction to create it such that it fails to verify after ‘x’ blocks; otherwise it should only get ejected if it could not fit in the mempool any longer — and for that it is dependent on how they are sorted, which today for free transactions is actually based on the one with the smaller hash; and since one could ad an arbitrary random attribute and keep creating transactions to get a smaller hash; it is already ordered by proof of work today for free transactions; it’s just not widely known.

So in conclusion; while I like making the MemPool it’s own class; I’m not really that enthusiastic about removing free transactions just for being old.

Sometimes someone might want to cancel their transaction that was free if it is getting old and it is tying up one of their utxo’s. For that I think it would be nice if they could sign a transaction requesting one of their previously free transactions be removed from mempools, and that be relayed around the network while nodes remove the transaction. That would allow them to send a new transaction with a fee and untie their utxo. That is the only real value I see to this business of removing an old free transaction (for the case it is tying up someone’s old utxo from being used again).

Comment (vncoelho) [reply neo/Ledger/MemPool.cs]: Hi, @jsolman, please, let's not even consider proof-of-work here. There is no reason for that.
Several other strategies can be applied, I prefer even use Learning techniques than using and PoW concept.

Previously, I had lost this part: "otherwise it should only get ejected if it could not fit in the mempool any longer"
It is important to take that is mind.

Great insights. I agree.
But I believe that something can still be designed, even with the idea of "address" signing the timestamp when making the request. But let the time tell the needs. heauheauae

**Comment by shargon** [comment]
I updated it to use two pools, note this:

- For query i search first in free pool (usually will be the more popular)
- For GetMemoryPool is used first paid pool
- I ported this line https://github.com/neo-project/neo/pull/392/files#diff-b95eb330cd65618c912779c81831051eR398 but i not sure why we need to clean the pools?

File: neo/Ledger/Blockchain.cs [hunk 11]
```diff
@@ -364,14 +380,24 @@ private RelayResultReason OnNewTransaction(Transaction transaction)
         private void OnPersistCompleted(Block block)
         {
             block_cache.Remove(block.Hash);
-            foreach (Transaction tx in block.Transactions)
-                mem_pool.TryRemove(tx.Hash, out _);
-            foreach (Transaction tx in mem_pool
+
+            foreach (var tx in block.Transactions)
+            {
+                if (!mem_pool_free.TryRemove(tx.Hash, out _))
```
Comment (vncoelho) [inline neo/Ledger/Blockchain.cs]: Maybe a verification can avoid Try to Remove from free if tx has fees.
something like tx.fees, right?

Comment (shargon) [reply neo/Ledger/Blockchain.cs]: You are right :)

Comment (vncoelho) [reply neo/Ledger/Blockchain.cs]: Thanks, myy friend :dagger:

**Comment by erikzhang** [comment]
> but i not sure why we need to clean the pools?

This is because all transactions in the memory pool need to be revalidated after the block is persisted.

**Comment by erikzhang** [comment]
I think we can put the two `Dictionary` in the class `MemPool`.

https://github.com/neo-project/neo/blob/9eb7a4a36ca7bfbdd2bd3614119822d065e77dba/neo/Ledger/MemPool.cs#L25

**Comment by shargon** [comment]
We can discuss this numbers https://github.com/neo-project/neo/pull/392/files#diff-b95eb330cd65618c912779c81831051eR121

**Comment by erikzhang** [comment]
> We can discuss this numbers ...

Maybe these two `ConcurrentDictionary` can share one maximum size. When the count of the two is added together to reach the maximum size, the transactions are preferentially removed from free.

**Comment by shargon** [comment]
Maybe we can change the threshold now?

**Comment by erikzhang** [comment]
The `threshold` (minimum fee for "free" transactions) should be a global constant. And every nodes should have the same value.

**Comment by shargon** [comment]
This constant should be in the config file. Or hard-coded?

**Comment by erikzhang** [comment]
In `protocol.json`?

**Comment by erikzhang** [comment]
What about setting `threshold` to `0.00999999`, so that any transaction with a fee below 0.01 gas should be considered as a low priority.

**Comment by shargon** [comment]
Maybe we can start with `0.000999999` for decrease the impact?

**Comment by shargon** [comment]
Note that this config should be replicated to:

https://github.com/neo-project/neo-cli/blob/master/neo-cli/protocol.testnet.json
https://github.com/neo-project/neo-cli/blob/master/neo-cli/protocol.json
https://github.com/neo-project/neo-cli/blob/master/neo-cli/protocol.mainnet.json

**Comment by shargon** [comment]
We can speed up this PR if you think that is fine?

File: neo/Network/P2P/Payloads/Transaction.cs [hunk 12]
```diff
@@ -49,6 +49,8 @@ public UInt256 Hash
 
         InventoryType IInventory.InventoryType => InventoryType.TX;
 
+        public bool IsLowPriority => NetworkFee < Settings.Default.LowPriorityThreshold;
```
Comment (belane) [inline neo/Network/P2P/Payloads/Transaction.cs]: what do you think if we 
`public bool IsLowPriority => NetworkFee < Settings.Default.LowPriorityThreshold || Type == TransactionType.ClaimTransaction;`
and treat all claims as low priority?

Comment (erikzhang) [reply neo/Network/P2P/Payloads/Transaction.cs]: It is a good idea.

```suggestion
        public bool IsLowPriority => Type == TransactionType.ClaimTransaction || NetworkFee < Settings.Default.LowPriorityThreshold;
```

File: neo/Network/P2P/Payloads/Transaction.cs [hunk 13]
```diff
@@ -49,6 +49,8 @@ public UInt256 Hash
 
         InventoryType IInventory.InventoryType => InventoryType.TX;
 
+        public bool IsLowPriority => Type == TransactionType.ClaimTransaction || NetworkFee < Settings.Default.LowPriorityThreshold;
```
Comment (shargon) [inline neo/Network/P2P/Payloads/Transaction.cs]: Good point!

File: neo/Ledger/MemoryPool.cs [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+            else
+            {
+                pool = _mem_pool_fee;
+            }
+
+            pool.TryAdd(hash, new PoolItem(tx));
+
+            if (Count > Capacity)
+            {
+                RemoveOldest(_mem_pool_free, DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 20));
```
Comment (jsolman) [inline neo/Ledger/MemoryPool.cs]: This basically will always do nothing, because after each block is persisted it re-adds the transactions and they get restamped with a new timestamp. I'm working on some major adjustments to this code that will improve performance, fix priority inversion that happens when adding transactions back to the chain after persist, and preserve the timestamp of the transaction across blocks.
